### ISSUE TITLE
Add t.before(fn), t.after(fn), t.beforeEach(fn), t.afterEach(fn)

### ISSUE THREAD

**Issue by isaacs** [tapjs/tapjs#138]
- `Test.before(function (t) {...})`: Run the function before any other subtests.  This is effectively just an alias for `t.test()`, but it only takes a function (name is always `"before"` or something).
- `Test.beforeEach(function (t) { ... })`: Alias for `t.test()` but runs before any subsequent test blocks in the section.
- `Test.after(function (t) { ... })`: Like tap v0.x's `t.tearDown(fn)` method.  Child test that runs after all other children have exited.  Not run in the case of bailouts.
- `Test.afterEach(function (t) { ... })`: Alias for `t.test()` but runs after any subsequent test block in the section.

**Comment by isaacs** [comment]
Just looked more closely at Mocha's use of these things.  The current `t.mochaGlobals` is not nearly mocha-y enough.  let's just do the `t.{before,after}{,Each}` stuff, and then split out the mochafication in another issue.

**Comment by themphill** [comment]
I'd love to see this feature land. Any idea on a timeline? I'd also be happy to help, if you prefer.

**Comment by isaacs** [comment]
No firm timeline at this time.

`beforeEach` and `afterEach` take a function. That function is called with a callback which they execute when they're finished.  Child tests are deferred until the callback is fired.

Unclear whether the before/afterEach handler should also get the test object.

If you want to add them, then that's fine.  Make sure to add lots of tests.  These can be added by putting a bunch of test scripts showing various beforeEach/afterEach behavior (multiple functions, verify that they run in order, etc etc) in `test/test/<etc>.js` and then running `node scripts/generate-test-test.js test/test/*.js` to generate the output once you have it doing the right thing how you want.  Also, they'll need docs.

**Comment by thewildpendulum** [comment]
Can do. I'll take a crack at it this week.

**Comment by ronkorving** [comment]
Any updates on this? For me this would be the killer feature to switch from node-tape to node-tap.

**Comment by micnic** [comment]
+1 for this idea
I would also add: `tap.before(function (end) {...})` and `tap.after(function (end) {...})` to run a function before and after all tests, usage:

``` js
tap.before(function (end) {
    setTimeout(function () {
        // do something async
        end();
    }, 1000);
});
```

**Comment by thewildpendulum** [comment]
The project I had intended to implement this feature for got sidelined for a while. I'm trying to find time to get it done in the next couple weeks, but I can't say with any certainty that it will happen. Anyone else interested in doing it sooner themselves is more than welcome to.

**Comment by isaacs** [comment]
API bikeshedding time!

Here are two options, with some pros and cons listed:
- https://gist.github.com/isaacs/3155c471014b5fdd19d3
- https://gist.github.com/isaacs/0950082cc78b4b9f7e35

Please let me know what you think.

**Comment by micnic** [comment]
I like the first option as it's simpler to implement and is similar to the the current API.

Beside this, I am wondering if these methods purpose is to do assertions, in my tests I would use them only to do some preparations for the actual test without any assestion.

**Comment by rmg** [comment]
I like the first option, except for one thing that I've always found annoying about mocha: assertions inside beforeEach/afterEach.

I haven't thought it through fully, and I really hate larger APIs, but I'm wondering about picking both options and calling the first option `preTest`/`postTest` and the second option `beforeEach`/`afterEach`.

**Comment by jamestalmage** [comment]
The ability to do assertions in `afterEach` can be useful. Sinons `mock.verify ()` is a good candidate.

I tend to agree that the first one looks way better.

**Comment by isaacs** [comment]
@micnic 

> if these methods purpose is to do assertions, in my tests I would use them only to do some preparations for the actual test without any assestion.

It's totally fine to have "tests" that don't do any assertions.  For example, this would be fine:

``` javascript
t.beforeEach(function (t) {
  setUpDatabase()
  deleteFixtures()
  t.end()
})
```

It's mostly just a question of which type of flow control API is best here.  Do you want to call `cb()` when you're finished, or `t.end()`?  And, should it be printed out as a test block, or just kept silent?

**Comment by isaacs** [comment]
Another option (which is an even weirder approach) is to use the callback style, but call in the context of the test object being before/after-ed.  So, in that case, you could do `this.equal(a,b)` in the before/after each function, but having a test object not called `t` kind of bugs me.  We don't rely on this-context in tap's user api surface at all, so it felt kind of like yet another weird point of friction.

**Comment by jamestalmage** [comment]
What if you suppressed the test block output unless there was a failure?

**Comment by jamestalmage** [comment]
I really think the API should stay consistent. Changing things for `beforeEach` would just be confusing.

**Comment by isaacs** [comment]
> What if you suppressed the test block output unless there was a failure?

I thought about that!  But it's a bit wonky implementation-wise (because now it's not going through the "normal" child-test code path), and also then it's kind of a step away from having a clear mostly-deterministic consistency between the test script and the TAP output.  For example, if you get a plan error because you added a beforeEach (which is a test), but that test doesn't show up in the TAP output, then it's pretty strange and hard to debug.

**Comment by isaacs** [comment]
As I play with this a bit more, I'm leaning more towards the callback route.  Here's why:
1. It doesn't mess with `t.plan()`, or create child tests unexpectedly, which is action at a distance a bit too spooky for my liking.
2. Even though the API is a bit out of step with the rest of node-tap, it's _not_ out of step with Node or JavaScript, and easy enough to solve with documentation.
3. It's less noisy.
4. The wonky implementation is already done and working, so whatever.  And I got it handling throws without any problem, so nbd there.

**Comment by Twipped** [comment]
:+1: to not altering t.plan or creating child tests.
:-1: to using a done callback instead of t.end(). The api should match.

Were you considering a before() and after() to wrap subtest runs? i know you can do t.on(end) to do post run stuff, but pre-run behavior always means wrapping the whole test set in a callback and it feels ugly. It would be nice to have a cleaner api for that

**Comment by isaacs** [comment]
@ChiperSoft I agree that it's better for the API to match.  But more importantly, the API should match the output.

If it's not altering t.plan or creating child tests, then `t.end()` is not the correct API surface.

**Comment by micnic** [comment]
@isaacs In general I am for silent output for these methods as they are not expected to do any assertion, and as you say that the API should match the output, in my opinion, we should choose the callback way to avoid any unexpected use of the methods.

**I vote for the callback approach!**

**Comment by ghost** [comment]
I vote for `no`. The economy of using `test()` for serial execution blocks is very simple and nice. I think a section in the docs about how to beforeEach/afterEach using `test()` would be just as good without adding more redundant clutter to the API.

**Comment by isaacs** [comment]
Landed callback-style on 5.2.0.

**Comment by pke** [comment]
How is this supposed to be used? I want to create a specific time for each of my tests, where would I save the generated time? Does `beforeEach` get some context where I could save the `startTime` I need?
